### ISSUE TITLE
FEMP and FEServer doesn't perform runtime Sanity Checks!

### ISSUE THREAD

**Issue by SapphireTactician** [eliatlarge/FEMultiPlayer-V2#94]
Issue type: `problem` `feature request` (*Both, deal with it*)

#### Description:

FEMP and FEServer lack runtime sanity checks and desperately needed to ensure that actions are valid and can be performed. FEMP and FEServer NEED to perform these, and I'm kinda scared by the fact that an online multiplayer game has managed to go this long between two groups of maintainers without it. @eliatlarge and @rayrobdod , with all due respect, I honestly should not have to be reporting this, this late into the project's life - which is why I didn't specifically check for it until now, after #91 and #92 caught my attention and caused me to examine the logs more carefully.

Currently, according to the log file and witnessing how the game behaves, it's clear that the game doesn't care about who ends the turn, meaning, if a player can spam End Turn, it'll view the second End Turn issued by the same player as the opponent forfeiting their turn, as cited here:

    [SEND]0 STARTGAME::
    [RCVE]1 COMMANDMESSAGE::Bartre move (4, 2):[WAIT]
    [SEND]1 COMMANDMESSAGE::Bartre move (4, 2):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]

This is the true heart of the turn skipping aspect of bugs #91 and #92 (*and Lord knows how many other bugs*), as there's no sanity check to verify if it's the turn player sending the command or not. Bug #93 actually helped make this easy to perform and verify. (*That being said, I propose we purposely postpone patching #93 for a while so that we can confirm fixes to this and the others once implemented*). All we had up to this point was a basic workaround by preventing the opponent from doing anything during their opponents turn, instead of ensuring that `ENDTURN` is only honored by the turn player. The subsequent `ENDTURN`(*s*), among other actions performed in #91 and #92  issued by the same player should've been rejected because it's no longer that player's turn.

I need some comfort food.

-------

#### Examples:

In the following log, Dart decides to troll Red Team Roy's team by exploiting the lack of Sanity Checking in FEMP and FEServer, tricking the game into thinking Red Roy's army ended turn cause reasons.

    [SEED] 0
    [RCVE]1 JOINLOBBY::net.fe.Player@1e713d6
    [SEND]1 JOINLOBBY::net.fe.Player@1e713d6
    [RCVE]1 JOINTEAM::1
    [SEND]1 JOINTEAM::1
    [RCVE]1 READYMESSAGE::
    [SEND]1 READYMESSAGE::
    [RCVE]2 JOINLOBBY::net.fe.Player@250a0c9c
    [SEND]2 JOINLOBBY::net.fe.Player@250a0c9c
    [RCVE]2 JOINTEAM::2
    [SEND]2 JOINTEAM::2
    [RCVE]2 READYMESSAGE::
    [SEND]2 READYMESSAGE::
    [SEND]0 STARTPICKING::
    [RCVE]2 PARTYMESSAGE::
        Roy Lv1
        Gilliam Lv1
        Lucius Lv1
        Kyle Lv1
        Knoll Lv1
        Matthew Lv1
        Tana Lv1
        Amelia Lv1
    [RCVE]1 PARTYMESSAGE::
        Roy Lv1
        Matthew Lv1
        Amelia Lv1
        Canas Lv1
        Artur Lv1
        Neimi Lv1
        Gilliam Lv1
        Dart Lv1
    [SEND]2 PARTYMESSAGE::
        Roy Lv1
        Gilliam Lv1
        Lucius Lv1
        Kyle Lv1
        Knoll Lv1
        Matthew Lv1
        Tana Lv1
        Amelia Lv1
    [SEND]1 PARTYMESSAGE::
        Roy Lv1
        Matthew Lv1
        Amelia Lv1
        Canas Lv1
        Artur Lv1
        Neimi Lv1
        Gilliam Lv1
        Dart Lv1
    [SEND]0 STARTGAME::
    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]

(*Continued*) Here, Dart gets a little too greedy and accidentally trolls himself by sending a 3rd End Turn, thus relinquishing control to Roy's army.

    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]

(*Continued*) Red Gilliam decides to troll Dart back for the lulz: (*Can you see I'm having too much fun with this by writing a story?*)

    [RCVE]2 COMMANDMESSAGE::Gilliam move (0, 0):[WAIT]
    [SEND]2 COMMANDMESSAGE::Gilliam move (0, 0):[WAIT]
    [RCVE]2 ENDTURN::
    [SEND]2 ENDTURN::
    [RCVE]2 ENDTURN::
    [SEND]2 ENDTURN::
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]

(*Continued*) Red Gilliam fouls up the cheese though by not being Sanic fast and is placed in the naughty corner

    [RCVE]2 COMMANDMESSAGE::Gilliam move (0, 0):[WAIT]
    [SEND]2 COMMANDMESSAGE::Gilliam move (0, 0):[WAIT]
    [RCVE]2 ENDTURN::
    [SEND]2 ENDTURN::
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]

(*Continued*) Dart resumes trolling:

    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Dart move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]

(*Continued*) Blue team going strong, Matthew wants in on the fun cause, well, this IS Matthew we're talking about:

    [RCVE]1 COMMANDMESSAGE::Matthew move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Matthew move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]

(*Continued*) Blue Team STILL trolling, Artur joins in to shit all over Red Roy's army:

    [RCVE]1 COMMANDMESSAGE::Artur move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Artur move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]

(*Continued*) Amelia doesn't like the mean game her friends are playing and ends turn normally to let Red Roy have a turn.

    [RCVE]1 COMMANDMESSAGE::Amelia move (0, 0):[WAIT]
    [SEND]1 COMMANDMESSAGE::Amelia move (0, 0):[WAIT]
    [RCVE]1 ENDTURN::
    [SEND]1 ENDTURN::
    [RCVE]2 COMMANDMESSAGE::[]
    [SEND]2 COMMANDMESSAGE::[]
    [RCVE]1 COMMANDMESSAGE::[]
    [SEND]1 COMMANDMESSAGE::[]

Blue Team lol's and walks away

    [RCVE]1 CHATMESSAGE::lolololol
    [SEND]1 CHATMESSAGE::lolololol
    [RCVE]1 QUITMESSAGE::
    [SEND]1 QUITMESSAGE::
    [SEND]ENDGAME:: winner = 2
    [RCVE]2 QUITMESSAGE::
    [SEND]2 QUITMESSAGE::

The End.

-------

#### Attachment:

[server_log68930794.log.txt](https://github.com/eliatlarge/FEMultiPlayer-V2/files/275253/server_log68930794.log.txt)

**Comment by SapphireTactician** [comment]
OMG, finally, we have Sanity Checks!!!!!!!!!! THANK YOU BOTH!!!!

Closing as `Fixed` as per #101 and #102 

Great job! @rayrobdod and @Purplemandown
